flake8_trio.py

```diff
@@ -498,108 +507,242 @@ def visit_Call(self, node: ast.Call):
+        if self.always_checkpoint is not None and (
+            not self.safe_decorator or self.yield_count
+        ):
+            self.error(
+                TRIO108, node.lineno, node.col_offset, "return", *self.always_checkpoint
+            )
+
         # avoid duplicate error messages
-        self.all_await = True
+        self.always_checkpoint = None
 
-    # disregard raise's in nested functions
+    # disregard checkpoints in nested function definitions
     def visit_FunctionDef(self, node: ast.FunctionDef):
-        outer = self.all_await
+        outer = self.get_state()
         self.generic_visit(node)
-        self.all_await = outer
+        self.set_state(outer)
 
     # checkpoint functions
-    def visit_Await(
-        self, node: Union[ast.Await, ast.AsyncFor, ast.AsyncWith, ast.Raise]
-    ):
+    def visit_Await(self, node: Union[ast.Await, ast.Raise]):
+        # the expression being awaited is not checkpointed
+        # so only set checkpoint after the await node
         self.generic_visit(node)
-        self.all_await = True
-
-    visit_AsyncFor = visit_Await
-    visit_AsyncWith = visit_Await
+        self.always_checkpoint = None
 
     # raising exception means we don't need to checkpoint so we can treat it as one
     visit_Raise = visit_Await
 
-    # valid checkpoint if there's valid checkpoints (or raise) in at least one of:
-    # (try or else) and all excepts
-    # finally
+    # checkpoint on one of enter and exit of with body
+    def visit_AsyncWith(self, node: ast.AsyncWith):
+        self.visit_nodes(node.items)
+        prebody_yield_count = self.yield_count
+        self.visit_nodes(node.body)
+        # if there was no yield in body that may unset a checkpoint on exit,
+        # treat this as a checkpoint
+        if prebody_yield_count == self.yield_count:
+            self.always_checkpoint = None
+
+    def visit_Yield(self, node: ast.Yield):
+        self.generic_visit(node)
+        self.yield_count += 1
+        if self.always_checkpoint is not None:
+            self.error(
+                TRIO108, node.lineno, node.col_offset, "yield", *self.always_checkpoint
+            )
+        self.always_checkpoint = ("yield", node.lineno)
+
+    # valid checkpoint if there's valid checkpoints (or raise) in:
+    # (try or else) and all excepts, or in finally
+
+    # try can jump into any except or into the finally at any point during it's execution
+    # so we need to make sure except & finally can handle a worst-case exception
     def visit_Try(self, node: ast.Try):
-        if self.all_await:
-            self.generic_visit(node)
-            return
+        # check worst case try exception
+        body_always_checkpoint = self.always_checkpoint
+        for n in (b for body in node.body for b in ast.walk(body)):
+            if isinstance(n, ast.Yield):
+                body_always_checkpoint = ("yield", n.lineno)
+                break
```

Comment:
Being able to combine the loop into one iter is nice to be able to break out easily, but there must be some way of doing the iteration cleaner.

Reply:
I'm still trying to come up with a solution to it with some friends for the fun of it, and got one that only needs 2 `for` but doesn't shortcut the innermost loop.
```python
for inner_node in itertools.chain(*(itertools.chain(ast.walk(body)) for body in node.body)):
```
But I wrote a wrapper for `ast.walk` that gave an easy-to-read solution that also is efficient.

Reply:
```python
for inner_node in itertools.chain.from_iterable(map(mywalk, node.body)):
```
is probably the best we came up with, but I suspect this is still less readable than the code currently in place, so I'll just leave it all as is.